packages/duckdb-wasm/src/bindings/runtime_browser.ts

```diff
@@ -91,6 +102,63 @@ export const BROWSER_RUNTIME: DuckDBRuntime & {
         }
     },
 
+    /** Prepare a file handle that could only be acquired aschronously */
+    async prepareDBFileHandle(dbPath: string, protocol: DuckDBDataProtocol): Promise<PreparedDBFileHandle[]> {
+        if (protocol === DuckDBDataProtocol.BROWSER_FSACCESS) {
+            const filePaths = [dbPath, `${dbPath}.wal`];
+            const prepare = async (path: string): Promise<PreparedDBFileHandle> => {
+                if (BROWSER_RUNTIME._files.has(path)) {
+                    return {
+                        path,
+                        handle: BROWSER_RUNTIME._files.get(path),
+                        fromCached: true,
+                    };
+                }
+                const opfsRoot = await navigator.storage.getDirectory();
+                let dirHandle: FileSystemDirectoryHandle = opfsRoot;
+                // check if mkdir -p is needed
+                const opfsPath = path.slice(OPFS_PREFIX_LEN);
+                let fileName = opfsPath;
+                if (PATH_SEP_REGEX.test(opfsPath)) {
+                    const folders = opfsPath.split(PATH_SEP_REGEX);
+                    fileName = folders.pop()!;
+                    if (!fileName) {
+                        throw new Error(`Invalid path ${path}`);
+                    }
+                    // mkdir -p
+                    for (const folder of folders) {
+                        dirHandle = await dirHandle.getDirectoryHandle(folder, { create: true });
+                    }
+                }
+                const fileHandle = await dirHandle.getFileHandle(fileName, { create: false }).catch(e => {
+                    if (e?.name === 'NotFoundError') {
+                        console.debug(`File ${path} does not exists yet, creating...`);
+                        return dirHandle.getFileHandle(fileName, { create: true });
+                    }
+                    throw e;
+                });
+                try {
+                    const handle = await fileHandle.createSyncAccessHandle();
```

Comment:
FileSystemSyncAccessHandles have strict locking requirements. If another web worker or tab has an open handle to this resource, this call will fail.

SQLite attempts to get around this with an automatic retry with exponential backoff. https://sqlite.org/forum/info/58a377083cd24a

It is also possible to implement a fully async awaitable lock with the Web Locks API: https://developer.mozilla.org/en-US/docs/Web/API/Web_Locks_API

A full implementation would likely use both approaches, since you don't know what other JS threads may be accessing the OPFS file that don't rely on web locks.

I think this is another reason to only provide opfs access through the `opfs://` protocol: I assume this access handle is released once the query is done? Or do these handles stay open for the lifetime of the db connection?

If the handles *are* automatically closed at the end of the query, then this would free up the lock to be used by other tabs / web workers. If you give users a sanctioned API to register `FileSystemSyncAccessHandles` through `registerFileHandle`, a user may shoot themselves in the foot by forgetting to call `duckdb.dropFile()` when they're done.

Reply:
Here, this PR does not consider other tabs.
It focuses on enabling the use of OPFS with WASM.

Reply:
It’s natural for a file to be locked when it’s being used by another tab. Therefore, I think it’s understandable that it results in an error.

The implementation of OPFS support for WASM and handling OPFS locking across tabs should be considered separately.

Reply:
Yes, the first step will be to get it merged.